### PR TITLE
[NIFI-774] Create a DeleteS3Object Processor

### PR DESCRIPTION
**Pull request by yu-iskw**
I'm sorry for the delay of my sending a new PR again. @danbress could you review it?

## JIRA
[[NIFI-774] Create a DeleteS3Object Processor - ASF JIRA](https://issues.apache.org/jira/browse/NIFI-774)

## Old PR
#72

### CODE REVIEW

**Comment by danbress** [comment]
Yu, this looks great.  Thanks again for sticking with it and contributing.  I made one comment inline.

I have one other comment that I'd like to get others feedback on:
This processor attempts to delete data from Amazon S3.  I could see there being a case for three scenarios, and three relationships out of the processor
1) the data is found in S3, and deleted(success)
2) the data is not found in S3, and therefore not deleted(not found)
3) there was some issue communicating with S3 and the object could not be deleted(failure)

Basically, you would be adding a "not found" relationship, and routing FlowFiles there when you they are not found in S3 before you attempt to delete them.  The reason being, you allow the user of your processor the opportunity to treat "not found" situations differently from "failure" situations.

I tried to find other cases like this in the NiFi processor marketplace, but I couldn't.  (https://github.com/apache/nifi/blob/master/nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/sqs/DeleteSQS.java) is the only "delete" processor I see, and it doesn't behave the way I am describing.

@markap14 @mcgilman @joewitt toughts

**Comment by yu-iskw** [comment]
@danbress thank you for the comment. That sounds good. It would be nice to add a property in order for users to handle if "not found" raises an error or not.

**Comment by yu-iskw** [comment]
@danbress could you review it? 
- Transfer `REL_NOT_FOUND` if not found target object on Amazon S3
- Add `REL_NOT_FOUND` to `DeleteS3Object`
- Redefine `relationships` on `DeleteS3Object` to append `REL_NOT_FOUND`
- Instead, get rid of `final` from `AbstractS3Processor`.`relationships`

**Comment by danbress** [comment]
Yu, looks good to me.  0.3.0 is being tested right now, as soon as it is tested and released I will merge these changes in.  Thanks again for your contribution.

**Comment by yu-iskw** [comment]
@danbress thank you for reviewing it. Sorry, one more thing, I fixed a typo and modified an error message. I look forward to merging it.

**Comment by taftster** [comment]
I'm not sure that checking for the existence of an existing object is that important.  At minimum, it's causing an extra call to the remote S3 service to determine if the object exists.  The S3 service itself will not return a different response code if the object previously existed.

What value is added by having the REL_NOT_FOUND relationship?  What use case is this trying to solve?

A request to delete an object, regardless of its previous existence, still results in the same remote state.  HTTP DELETE, being an idempotent method, doesn't really care if the object existed beforehand or not.

**Comment by yu-iskw** [comment]
@danbress @taftster I removed the process if the key exists or not. Could you review it?

File: nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/s3/DeleteS3Object.java [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+
+
+@SupportsBatching
+@SeeAlso({PutS3Object.class})
+@Tags({"Amazon", "S3", "AWS", "Archive", "Delete"})
+@CapabilityDescription("Deletes FlowFiles on an Amazon S3 Bucket. " +
+        "And the FlowFiles are checked if exists or not before deleting.")
+public class DeleteS3Object extends AbstractS3Processor {
+
+    public static final Relationship REL_NOT_FOUND = new Relationship.Builder().name("not found")
```
Comment (danbress) [inline nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/s3/DeleteS3Object.java]: Not needed anymore

File: nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/s3/DeleteS3Object.java [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+            .expressionLanguageSupported(true)
+            .required(false)
+            .build();
+
+    public static final List<PropertyDescriptor> properties = Collections.unmodifiableList(
+            Arrays.asList(KEY, BUCKET, ACCESS_KEY, SECRET_KEY, CREDENTAILS_FILE, REGION, TIMEOUT, VERSION_ID,
+                    FULL_CONTROL_USER_LIST, READ_USER_LIST, WRITE_USER_LIST, READ_ACL_LIST, WRITE_ACL_LIST, OWNER));
+
+    public static final Set<Relationship> relationships = Collections.unmodifiableSet(
+            new HashSet<>(Arrays.asList(REL_SUCCESS, REL_FAILURE, REL_NOT_FOUND)));
```
Comment (danbress) [inline nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/s3/DeleteS3Object.java]: REL_NOT_FOUND should not be in this list anymore

**Comment by danbress** [comment]
@yu-iskw Looks great!  Only thing I see right now is that the REL_NOT_FOUND relationship isn't needed anymore.  Thanks again for the contribution.  I will try and merge this in as soon as 0.3.0 is released.

**Comment by yu-iskw** [comment]
@danbress thank you for pointing out the easy mistake. I fixed it.

**Comment by yu-iskw** [comment]
Thank you for merging it!

**Pull request merged**